Userland/Applications/Calendar/EventManager.cpp

```diff
@@ -97,14 +97,82 @@ ErrorOr<Vector<Event>> EventManager::deserialize_events(JsonArray const& json)
     return result;
 }
 
+Core::DateTime EventManager::format_icalendar_vevent_datetime(String const& parameter)
+{
+    auto invalid_datetime = Core::DateTime::create(0);
+    if (parameter.to_byte_string().length() < 15)
+        return invalid_datetime;
+    auto formatted_string = String::formatted("{}-{}-{}T{}:{}:{}",
+        parameter.substring_from_byte_offset(0, 4).value(),
+        parameter.substring_from_byte_offset(4, 2).value(),
+        parameter.substring_from_byte_offset(6, 2).value(),
+        parameter.substring_from_byte_offset(9, 2).value(),
+        parameter.substring_from_byte_offset(11, 2).value(),
+        parameter.substring_from_byte_offset(13, 2).value());
+    if (formatted_string.is_error())
+        return invalid_datetime;
+    auto datetime = Core::DateTime::parse(DATE_FORMAT, formatted_string.value());
+    return datetime.has_value() ? datetime.value() : invalid_datetime;
+}
+
+ErrorOr<Vector<Event>> EventManager::parse_icalendar_vevents(ByteBuffer const& content)
+{
+    Event event;
+    Vector<Event> events;
+    ICalendarParserState state = ICalendarParserState::Idle;
+    StringBuilder builder;
+    TRY(builder.try_append((char const*)content.data(), content.size()));
+    auto lines = TRY(TRY(builder.to_string()).split('\n'));
+    for (size_t i = 0; i < lines.size(); i++) {
+        auto section = TRY(lines[i].split_limit(':', 2));
+        if (section.size() > 1) {
+            auto property = section[0];
+            auto parameter = TRY(section[1].trim_ascii_whitespace());
+            switch (state) {
+            case ICalendarParserState::InVEvent:
+                if (property == "DTSTART")
+                    event.start = format_icalendar_vevent_datetime(parameter);
+                if (property == "DTEND")
+                    event.end = format_icalendar_vevent_datetime(parameter);
+                if (property == "SUMMARY")
+                    event.summary = parameter;
+                if (property == "END" && parameter == "VEVENT") {
+                    if (event.start.year() && event.end.year())
+                        events.append(event);
+                    state = ICalendarParserState::Idle;
+                }
+                break;
+            case ICalendarParserState::Idle:
+                if (property == "BEGIN" && parameter == "VEVENT")
+                    state = ICalendarParserState::InVEvent;
+                break;
+            default:
+                break;
+            }
+        }
+    }
+    return events;
+}
+
+ErrorOr<Vector<Event>> EventManager::parse_events(ByteBuffer const& content)
+{
+    // If content is iCalendar format, try to parse VEVENTs
+    if (content.size() > 15 && !memcmp(content.data(), "BEGIN:VCALENDAR", 15)) {
```

Comment:
Something like `if (content.span().starts_with("BEGIN:VCALENDAR"sv.bytes())` should work and is a bit nicer